[addons] add support to new addon gui interface for list items

AlwinEsch opened the pull request:
## Description
<!--- Describe your change in detail -->
This changes at the basic support about list items to use on addon
window.

On old system are big memleaks present. This new use now shared pointers
on addon and on kodi.

Also shares Kodi his shared pointer of addon list items as void* to addon,
means one shared_ptr becomes generated with new and contains related shared list
item pointer. On every call back becomes this pointer readed and needed
shared pointer for kodi is present.

The destruct of them pointet shared mem pointer becomes done from Kodi during
destruct of addons list item.

During several tests works it very nice and no memleaks or blockages present.

<!--- Provide a general summary of your change in the Title above -->


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
Here one of my addon test codes:
```cpp
bool CMyWindow::OnInit()
{
  ClearList();

  AddListItem(std::make_shared<kodi::gui::CListItem>("Test 1 - direct"));

  kodi::gui::ListItemPtr item(new kodi::gui::CListItem("Test 2 - other"));
  item->SetPath("home/alwin");
  item->SetIconImage("home/alwin/1.png");
  AddListItem(item);

  for (unsigned int i = 2; i < 100; i++)
  {
    AddListItem(StringUtils::Format("Test %i", i));
  }

  fprintf(stderr, "-TEST: %s - %s\n", GetListItem(1)->GetLabel().c_str(), GetListItem(1)->GetIconImage().c_str());

  GetListItem(3)->SetPath("/home/alwin");
  fprintf(stderr, "-TEST: %s - %s\n", GetListItem(3)->GetLabel().c_str(), GetListItem(3)->GetPath().c_str());
}
```

Rechi commented on xbmc/addons/interfaces/GUI/ListItem.cpp:
```diff
@@ -34,12 +34,39 @@ extern "C"
+public:
+  CAddonFileItem()
+  {
+    fprintf(stderr, "file item count: %i; %s\n", ++fileitemcnt, __PRETTY_FUNCTION__);
```
`__PRETTY_FUNCTION__` can't be used on windows, replace it with `__FUNCTION__`
same is true for the other occurrences of `__PRETTY_FUNCTION__`

ksooo commented on xbmc/addons/interfaces/GUI/ListItem.cpp:
```diff
@@ -87,5 +112,236 @@ void Interface_GUIListItem::destroy(void* kodiBase, void* handle)
+    return nullptr;
+  }
+
+  if (item->get() == nullptr)
```
You could just do a <code>if (!item)</code> here and at several other places.

AlwinEsch replied:
Not in this case with `->`, as alternative can be only `if (!*item)`.

ksooo replied:
Oops, I actually ment <code>!*item</code>.

ksooo commented on xbmc/addons/interfaces/GUI/ListItem.cpp:
```diff
@@ -87,5 +112,236 @@ void Interface_GUIListItem::destroy(void* kodiBase, void* handle)
+
+  char* ret;
+  Interface_GUIGeneral::lock();
+  ret = strdup(item->get()->GetLabel().c_str());
```
I would use <code>*item->Foo()</code> instead of <code>item->get()->Foo()</code> here and at several other places.

AlwinEsch replied:
Does not work and bring:
```
/home/alwin/Development/kodi/xbmc/addons/interfaces/GUI/ListItem.cpp: In static member function ‘static char* ADDON::Interface_GUIListItem::get_label(void*, void*)’:
/home/alwin/Development/kodi/xbmc/addons/interfaces/GUI/ListItem.cpp:134:23: error: ‘CFileItemPtr {aka class std::shared_ptr<CFileItem>}’ has no member named ‘GetLabel’
   ret = strdup(*item->GetLabel().c_str());
```
Thats why I leaved also as `if (item->get() == nullptr)` before.

ksooo replied:
Oops. Use <code (*item) ->foo()</code>, please.

ksooo commented on xbmc/addons/interfaces/GUI/ListItem.cpp:
```diff
@@ -87,5 +112,236 @@ void Interface_GUIListItem::destroy(void* kodiBase, void* handle)
+  return ret;
+}
+
+void Interface_GUIListItem::set_icon_image(void* kodiBase, void* handle, const char *label)
```
Copy/paste error? Parameter name <code>label</code> does not match what it actually contains. Should read 'image' or the like. 

I've seen this name mismatch in more setters in this source file. All names should be fixed.

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -278,6 +281,63 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+  }
+
+  Interface_GUIGeneral::lock();
+  static_cast<CGUIAddonWindow*>(handle)->ClearList();
```
You should use variable 'pAddonWindow' here as it already contains the value of <code>static_cast<CGUIAddonWindow*>(handle)</code>.

Same glitch is present in other functions in this source file.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/ListItem.h:
```diff
@@ -22,22 +22,44 @@
-    virtual ~CAddonGUIControlBase() = default;
+      m_Window(window)
+    {
+      /// TEMPORARY PRESENT FOT TEST PURPOSE ONLY !!!
```
May I suggest that we remove all test code before merge?

AlwinEsch replied:
:-D

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/ListItem.h:
```diff
@@ -22,22 +22,44 @@
-  class CAddonGUIControlBase
+  class CWindow;
+
+  class CAddonGUIControlBase : public std::enable_shared_from_this<CAddonGUIControlBase>
```
Do you really need <code>shared_from_this</code>? Cannot find any reference...

AlwinEsch replied:
Becomes removed, has tested here to use them, but found on class constructor not usable and brings a `std::bad_weak_ptr` exception.

ksooo reviewed (changes requested):
A little more love is needed, imo.

AlwinEsch commented:
Is updated.

AlwinEsch committed:
[addons] add support to new addon gui interface for list items

This changes at the basic support about list items to use on addon
window.

On old system are big memleaks present. This new use now shared pointers
on addon and on kodi.

Also shares Kodi his shared pointer of addon list items as void* to addon,
means one shared_ptr becomes generated with new and contains related shared list
item pointer. On every call back becomes this pointer readed and needed
shared pointer for kodi is present.

The destruct of them pointet shared mem pointer becomes done from Kodi during
destruct of addons list item.

During several tests works it very nice and no memleaks or blockages present.

ksooo commented:
Fine by me now. It's up to you whether you want to get rid of the item->get() calls.

ksooo commented:
jenkins build this please
